### DISCUSSION TITLE
Confusion regarding lifetimes

### DISCUSSION THREAD

**Discussion by JuanDiegoMontoya** [jrouwe/JoltPhysics#1369] [Q&A] [answered]
(I'm using v5.2.0)

I have some code that creates a sphere shape, then uses it in creating one or more bodies. There are two ways of creating a sphere shape as far as I can tell.

1:
```cpp
auto sphereSettings = JPH::SphereShapeSettings(1);
sphereSettings.SetEmbedded();
auto sphere = sphereSettings.Create().Get(); // `sphere` is Ref<Shape>
```

2:
```cpp
auto sphere = JPH::SphereShape(1); // `sphere` is on the stack
```

However, when `sphere` goes out of scope in the second version, an assert in [~RefTarget()](https://github.com/jrouwe/JoltPhysics/blob/master/Jolt/Core/Reference.h#L40) is tripped. It doesn't seem to matter whether I call `sphere.SetEmbedded();`.

The assert no longer trips if I `new` the `SphereShape` (including if it's wrapped in a `JPH::Ref`).

I guess I have two main questions:

1. Is it possible to create a shape on the stack?
    - The docs say `Simple shapes like spheres and boxes can be constructed immediately by simply new-ing them.`. which implies that they cannot live on the stack. It's not clear why that is.
2. When am I supposed to call `SetEmbedded`? What happens if I call `SetEmbedded` unnecessarily or if I forget to call it (I predict that the former is a leak while the latter is a crash)?
    - The docs say `Mark this class as embedded, this means the type can be used in a compound or constructed on the stack.`.
    - The hello world example has a comment next to a usage of `SetEmbedded`: `A ref counted object on the stack (base class RefTarget) should be marked as such to prevent it from being freed when its reference count goes to 0.`
    - It's not clear to me why it wouldn't work for shapes as they are refcounted objects.

**Comment by jrouwe** [comment, marked as answer]
You cannot create a shape on the stack if you want to attach that shape to a body. The body doesn't make a copy of the shape, it takes a reference. So unless you create and destroy a body within the scope where `sphere` exists, you'll get crashes.

That said, you can do things like:

```
SphereShape sphere(1.0f);
sphere.SetEmbedded();
physics_system->GetNarrowPhaseQuery().CollideShape(&sphere, ...);
```

because then `sphere` lives long enough for the narrow phase query to be done with it and it is more efficient than:

```
SphereShape *sphere = new SphereShape(1.0f);
physics_system->GetNarrowPhaseQuery().CollideShape(sphere, ...);
delete sphere;
```

**Reply by JuanDiegoMontoya** [reply]
I see now. I had a broken preconception of what shapes did internally, but this makes sense.

Thanks!
